### PR TITLE
First set of test for SchemaCompareDialog

### PR DESCRIPTION
**Pull request by SakshiS-harma**
Added one test for a util method as well.

### CODE REVIEW

File: extensions/schema-compare/src/test/testSchemaCompareMainWindow.ts [hunk 1]
```diff
@@ -69,6 +69,19 @@ export class SchemaCompareMainWindowTest extends SchemaCompareMainWindow {
 			result = true;
 		}
 
+		if (result === false) {
```
Comment (Charles-Gagnon) [inline extensions/schema-compare/src/test/testSchemaCompareMainWindow.ts:72]: Why not just remove the initial if and then use should to actually do the verification here?

File: extensions/schema-compare/src/dialogs/schemaCompareDialog.ts [hunk 2]
```diff
@@ -53,15 +47,17 @@ export class SchemaCompareDialog {
 	private targetDbEditable: string;
 	private previousSource: mssql.SchemaCompareEndpointInfo;
 	private previousTarget: mssql.SchemaCompareEndpointInfo;
+	private initDialogComplete: Deferred<void>;
```
Comment (kisantia) [inline extensions/schema-compare/src/dialogs/schemaCompareDialog.ts:50]: thanks for adding this! Can you also add this to the options dialog?

File: extensions/schema-compare/src/test/schemaCompareDialog.test.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		let dialog = new SchemaCompareDialog(schemaCompareResult);
+		await dialog.openDialog();
+
+		should(dialog.dialog.title).equal('Schema Compare');
+		should(dialog.dialog.okButton.label).equal('OK');
+		should(dialog.dialog.okButton.enabled).equal(false); // Should be false when open
+	});
+
+	it('Simulate ok button- with both endpoints set to dacpac', async function (): Promise<void> {
+		//let getCurrentConnectionResult: azdata.connection.ConnectionProfile[] = {...mockConnectionProfile};
```
Comment (kisantia) [inline extensions/schema-compare/src/test/schemaCompareDialog.test.ts:48]: remove commented code

File: extensions/schema-compare/src/test/schemaCompareDialog.test.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		schemaCompareResult.sourceEndpointInfo = setDacpacEndpointInfo(mocksource);
+		schemaCompareResult.targetEndpointInfo = setDacpacEndpointInfo(mocktarget);
+
+		let dialog = new SchemaCompareDialog(schemaCompareResult);
+		await dialog.openDialog();
+
+		await dialog.execute();
+
+		// Confirm that ok button got clicked
+		should(schemaCompareResult.verifyButtonsState( {compareButtonState: true, optionsButtonState: true, switchButtonState: true,
```
Comment (kisantia) [inline extensions/schema-compare/src/test/schemaCompareDialog.test.ts:61]: is the state of one or more of the buttons supposed to change after Ok is clicked? If yes, then please add a check for the button states before the dialog is opened to verify that the state changed as expected

File: extensions/schema-compare/src/dialogs/schemaCompareDialog.ts (hunk 2 above)
Comment (Charles-Gagnon) [reply extensions/schema-compare/src/dialogs/schemaCompareDialog.ts:50]: FYI - this is probably useful to have as a base abstract class things can extend from so it's easy to add to dialogs as you see fit. See https://github.com/microsoft/azuredatastudio/blob/main/extensions/arc/src/ui/components/initializingComponent.ts as an example

File: extensions/schema-compare/src/test/testSchemaCompareMainWindow.ts (hunk 1 above)
Comment (Benjin) [reply extensions/schema-compare/src/test/testSchemaCompareMainWindow.ts:72]: Or just return true immediately, leaving the logging if it never returns.

File: extensions/schema-compare/src/test/schemaCompareDialog.test.ts (hunk 4 above)
Comment (Benjin) [inline extensions/schema-compare/src/test/schemaCompareDialog.test.ts:61]: Could you break the param object into multiple lines for readability, please?

Comment (SakshiS-harma) [reply extensions/schema-compare/src/test/schemaCompareDialog.test.ts:61]: For this test case, nothing is supposed to change- the button states remain the same, since both source and destination are already populated.
I will look into adding more test cases where one or both are empty and get populated after this dialog opens. But for now, I will leave it for next PR.

File: extensions/schema-compare/src/test/schemaCompareDialog.test.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		mockExtensionContext = TypeMoq.Mock.ofType<vscode.ExtensionContext>();
+		mockExtensionContext.setup(x => x.extensionPath).returns(() => '');
+	});
+
+	it('Should be correct when created.', async function (): Promise<void> {
+		let schemaCompareResult = new SchemaCompareMainWindow(testContext.apiWrapper.object, undefined, mockExtensionContext.object);
+		let dialog = new SchemaCompareDialog(schemaCompareResult);
+		await dialog.openDialog();
+
+		should(dialog.dialog.title).equal('Schema Compare');
```
[the remark this answers is not in the document]
Comment (SakshiS-harma) [reply extensions/schema-compare/src/test/schemaCompareDialog.test.ts]: Updated it appropriately.

File: extensions/schema-compare/src/test/testSchemaCompareMainWindow.ts (hunk 1 above)
Comment (SakshiS-harma) [reply extensions/schema-compare/src/test/testSchemaCompareMainWindow.ts]: I am keeping the actual verification to the test cases only. 
Updated it based on Benjin's comment.

File: extensions/schema-compare/src/dialogs/schemaCompareDialog.ts (hunk 2 above)
Comment (SakshiS-harma) [reply extensions/schema-compare/src/dialogs/schemaCompareDialog.ts:50]: That is a great example, Charles. 
I will update the options dialog in another PR and will introduce the base abstract class thing there (when it will actually be needed to be done at two places). So for now, kept it as-is.

File: extensions/schema-compare/src/test/schemaCompareDialog.test.ts (hunk 3 above)
Comment (SakshiS-harma) [reply extensions/schema-compare/src/test/schemaCompareDialog.test.ts]: Done.

File: extensions/schema-compare/src/test/schemaCompareDialog.test.ts (hunk 4 above)
Comment (SakshiS-harma) [reply extensions/schema-compare/src/test/schemaCompareDialog.test.ts]: Done.

File: extensions/schema-compare/src/test/testSchemaCompareMainWindow.ts [hunk 6]
```diff
@@ -66,9 +64,20 @@ export class SchemaCompareMainWindowTest extends SchemaCompareMainWindow {
 			this.selectTargetButton.enabled === buttonState.selectTargetButtonState &&
 			this.generateScriptButton.enabled === buttonState.generateScriptButtonState &&
 			this.applyButton.enabled === buttonState.applyButtonState) {
-			result = true;
+			return true;
 		}
 
-		return result;
+		console.log('CompareButton: (Actual) ', this.compareButton.enabled, '(Expected) ', buttonState.compareButtonState);
```
Comment (kisantia) [inline extensions/schema-compare/src/test/testSchemaCompareMainWindow.ts:70]: it's a little strange that these are console.logs instead of should

File: extensions/schema-compare/src/utils.ts [hunk 7]
```diff
@@ -84,7 +85,7 @@ function connectionInfoToConnectionProfile(details: azdata.ConnectionInfo): azda
 	};
 }
 
-export async function verifyConnectionAndGetOwnerUri(endpoint: mssql.SchemaCompareEndpointInfo, caller: string, apiWrapper: ApiWrapper): Promise<string> {
+export async function verifyConnectionAndGetOwnerUri(endpoint: mssql.SchemaCompareEndpointInfo, caller: string, apiWrapper: ApiWrapper): Promise<string | undefined> {
```
Comment (Benjin) [inline extensions/schema-compare/src/utils.ts:88]: Would it be better to throw an error rather than have split logic above this for `undefined`?  No idea myself.

File: extensions/schema-compare/src/dialogs/schemaCompareDialog.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (SakshiS-harma) [reply extensions/schema-compare/src/dialogs/schemaCompareDialog.ts:50]: The promise needs to end later , hence the use of deferred promise. There might be a better way but not that I am aware of for now.

File: extensions/schema-compare/src/test/testSchemaCompareMainWindow.ts (hunk 6 above)
Comment (SakshiS-harma) [reply extensions/schema-compare/src/test/testSchemaCompareMainWindow.ts]: Fixed it to use "should" and also printing all the values when it fails.

File: extensions/schema-compare/src/test/utils.test.ts [hunk 8]
```diff
@@ -8,7 +8,11 @@ import * as azdata from 'azdata';
 import * as mssql from '../../../mssql';
 import * as loc from '../localizedConstants';
 import * as TypeMoq from 'typemoq';
-import {getEndpointName, verifyConnectionAndGetOwnerUri } from '../utils';
+import * as path from 'path';
+import * as uuid from 'uuid';
+import * as os from 'os';
+import { promises as fs } from 'fs';
+import {getEndpointName, verifyConnectionAndGetOwnerUri, exists } from '../utils';
```
[the remark this answers is not in the document]
Comment (SakshiS-harma) [reply extensions/schema-compare/src/test/utils.test.ts]: Done.

File: extensions/schema-compare/src/utils.ts (hunk 7 above)
Comment (SakshiS-harma) [reply extensions/schema-compare/src/utils.ts:88]: Actually, the logic already exists to handle both the cases. I just realized that the last time when I worked on this file, I missed this change due to some merge conflicts, hence introduced this change back here.
